flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/TaskBalancedPreferredSlotSharingStrategy.java

```diff
@@ -0,0 +1,344 @@
+            final int subtaskIndex = sev.getId().getSubtaskIndex();
+            return coLocationGroupMap.get(jobVertexID).getLocationConstraint(subtaskIndex);
+        }
+
+        private int getSlotRoundRobinIndex(
+                final int jobVertexParallelism, SlotSharingGroup slotSharingGroup) {
+            final boolean maxParallel = isMaxParallelism(jobVertexParallelism, slotSharingGroup);
+            return maxParallel ? 0 : slotSharingGroupIndexMap.getOrDefault(slotSharingGroup, 0);
+        }
+
+        private void updateSlotRoundRobinIndexIfNeeded(
+                final int jobVertexParallelism,
+                final SlotSharingGroup slotSharingGroup,
+                final int nextIndex) {
+            if (!isMaxParallelism(jobVertexParallelism, slotSharingGroup)) {
+                slotSharingGroupIndexMap.put(slotSharingGroup, nextIndex);
+            }
+        }
+
+        private boolean isMaxParallelism(
+                final int jobVertexParallelism, final SlotSharingGroup slotSharingGroup) {
+            final List<ExecutionSlotSharingGroup> executionSlotSharingGroups =
+                    paralleledExecutionSlotSharingGroupsMap.get(slotSharingGroup);
+            return jobVertexParallelism == executionSlotSharingGroups.size();
+        }
+
+        private int getLeastUtilizeSlotIndex(
+                final List<ExecutionSlotSharingGroup> executionSlotSharingGroups,
+                @Nullable final SchedulingExecutionVertex executionVertex) {
+            int indexWithLeastExecutionVertices = 0;
+            int leastExecutionVertices = Integer.MAX_VALUE;
+            for (int index = 0; index < executionSlotSharingGroups.size(); index++) {
+                final ExecutionSlotSharingGroup executionSlotSharingGroup =
+                        executionSlotSharingGroups.get(index);
+                final int executionVertices =
+                        executionSlotSharingGroup.getExecutionVertexIds().size();
+                if (leastExecutionVertices > executionVertices
+                        && (Objects.isNull(executionVertex)
+                                || allocatable(executionSlotSharingGroup, executionVertex))) {
+                    indexWithLeastExecutionVertices = index;
+                    leastExecutionVertices = executionVertices;
+                }
+            }
+            return indexWithLeastExecutionVertices;
+        }
+
+        private boolean allocatable(
+                final ExecutionSlotSharingGroup executionSlotSharingGroup,
+                @Nonnull SchedulingExecutionVertex executionVertex) {
+            final ExecutionVertexID executionVertexId = executionVertex.getId();
+            final JobVertexID jobVertexId = executionVertexId.getJobVertexId();
+            final Set<ExecutionVertexID> existedExecutionVertexIds =
+                    executionSlotSharingGroup.getExecutionVertexIds();
+            final Set<JobVertexID> allocatedJobVertices =
```

Comment:
Hi RocMarshal! From my point of view, the process of building the JobVertexId set may will be time-consuming in double-layer nested loops.